include/nbl/builtin/hlsl/bxdf/common.hlsl

```diff
@@ -48,358 +139,976 @@ struct Basic
-    if (valid)
-    {
-      retval.TdotH = dot(interaction.T,H);
-      retval.BdotH = dot(interaction.B,H);
-    }
-    return valid;
-  }
-
-  float TdotH;
-  float BdotH;
+    using this_t = SIsotropicMicrofacetCache<T>;
+    using scalar_type = T;
+    using vector3_type = vector<scalar_type, 3>;
+    using matrix3x3_type = matrix<scalar_type, 3, 3>;
+
+    using ray_dir_info_type = ray_dir_info::SBasic<scalar_type>;
+    using isotropic_type = surface_interactions::SIsotropic<ray_dir_info_type>;
+    using sample_type = SLightSample<ray_dir_info_type>;
+
+    // always valid because its specialized for the reflective case
+    static this_t createForReflection(const scalar_type NdotV, const scalar_type NdotL, const scalar_type VdotL, NBL_REF_ARG(scalar_type) LplusV_rcpLen)
+    {
+        LplusV_rcpLen = rsqrt<scalar_type>(2.0 + 2.0 * VdotL);
+
+        this_t retval;
+
+        retval.VdotH = LplusV_rcpLen * VdotL + LplusV_rcpLen;
+        retval.LdotH = retval.VdotH;
+        retval.NdotH = (NdotL + NdotV) * LplusV_rcpLen;
+        retval.NdotH2 = retval.NdotH * retval.NdotH;
+
+        return retval;
+    }
+    static this_t createForReflection(const scalar_type NdotV, const scalar_type NdotL, const scalar_type VdotL)
+    {
+        float dummy;
+        return createForReflection(NdotV, NdotL, VdotL, dummy);
+    }
+    template<class ObserverRayDirInfo, class IncomingRayDirInfo NBL_FUNC_REQUIRES(ray_dir_info::Basic<ObserverRayDirInfo> && ray_dir_info::Basic<IncomingRayDirInfo>)
+    static this_t createForReflection(
+        NBL_CONST_REF_ARG(surface_interactions::SIsotropic<ObserverRayDirInfo>) interaction,
+        NBL_CONST_REF_ARG(SLightSample<IncomingRayDirInfo>) _sample)
+    {
+        return createForReflection(interaction.getNdotV(), _sample.getNdotL(), _sample.getVdotL());
+    }
+    // transmissive cases need to be checked if the path is valid before usage
+    static bool compute(
+        NBL_REF_ARG(this_t) retval,
+        const bool transmitted, NBL_CONST_REF_ARG(vector3_type) V, NBL_CONST_REF_ARG(vector3_type) L,
+        NBL_CONST_REF_ARG(vector3_type) N, const scalar_type NdotL, const scalar_type VdotL,
+        const scalar_type orientedEta, const scalar_type rcpOrientedEta, NBL_REF_ARG(vector3_type) H
+    )
+    {
+        // TODO: can we optimize?
+        H = computeMicrofacetNormal<vector3_type>(transmitted,V,L,orientedEta);
+        retval.NdotH = nbl::hlsl::dot<vector3_type>(N, H);
+
+        // not coming from the medium (reflected) OR
+        // exiting at the macro scale AND ( (not L outside the cone of possible directions given IoR with constraint VdotH*LdotH<0.0) OR (microfacet not facing toward the macrosurface, i.e. non heightfield profile of microsurface) )
+        const bool valid = !transmitted || (VdotL <= -min(orientedEta, rcpOrientedEta) && retval.NdotH > nbl::hlsl::numeric_limits<scalar_type>::min);
```

Comment:
in a setting where you're already provided `NdotL` and can compute `NdotV` (you have V and N) you can work out `transmission` from `isTransmissionPath(NdotV,NdotL)`

you can also have a `create(const vector_type N, const vector_type V, const vector_type L, REF_ARG(vector_type) H)` which can work out the transmission for itself.

Reply:
also that `>min` should be `>=`